### ISSUE TITLE
rPCA inconsistency?

### ISSUE THREAD

**Issue by Konrad0** [hyperspy/hyperspy#1858]
Hi all,

recently I may have spotted an inconsistency in Hyperspy's implementation of robust PCA. Subsequently I tried to narrow it down in an idealized test case and keep getting the same strange behavior. So the test case consists of a matrix ```A``` full of ones, except for one element, which is set to two. In my understanding, this should be decomposed into a matrix ```L``` filled with ones and a matrix ```S``` filled with zeros and a one in the element corresponding to the two in ```A```, such that approximately ```L+S = A```. I checked this with another implementation (```r_pca``` from https://github.com/dganguli/robust-pca), which gave the expected result, but Hyperspy disagreed completely. Is this a bug or am I using Hyperspy's rPCA routine incorrectly? I'm using Hyperspy version 1.3. Here's my code:

```python
import numpy as np

import importlib
importlib.import_module('r_pca')
from r_pca import R_pca

import hyperspy.api as hs

n=5
m=10
A = np.ones((n,m))
A[int(n/2),int(m/2)] = 2.0

rpca = R_pca(A)
L, S = rpca.fit(max_iter=1000, iter_print=100)

hsdata = hs.signals.Signal1D(A)
Lhs,Shs = hsdata.decomposition(algorithm='RPCA_GoDec', return_info=True, output_dimension=min(np.shape(A)), power=3)

print('Ganguli-Candes: L= ')
print(L)

print('Hyperspy: L= ')
print(Lhs)

print('Ganguli-Candes: S= ')
print(S)

print('Hyperspy: S= ')
print(Shs)

print('Ganguli-Candes: L+S= ')
print(L+S)

print('Hyperspy: L+S= ')
print(Lhs+Shs)

print('Done.')

import hyperspy
print('Hyperspy version '+str(hyperspy.__version__))
```

And here's the corresponding output:
```
iteration: 1, error: 18.36113064260148
iteration: 8, error: 1.3991493881412786e-06
WARNING:hyperspy.learn.rpca:Threshold 'lambda1' is set to default: 1 / sqrt(nsamples)
WARNING:hyperspy.learn.rpca:Convergence tolerance not specifed. Defaulting to 1e-3
WARNING:hyperspy.learn.rpca:Maximum iterations not specified. Defaulting to 1e3
Ganguli-Candes: L=
[[1.00010184 1.00010184 1.00010184 1.00010184 1.00010184 1.00019043
  1.00010184 1.00010184 1.00010184 1.00010184]
 [1.00010184 1.00010184 1.00010184 1.00010184 1.00010184 1.00019043
  1.00010184 1.00010184 1.00010184 1.00010184]
 [1.00031282 1.00031282 1.00031282 1.00031282 1.00031282 1.00040142
  1.00031282 1.00031282 1.00031282 1.00031282]
 [1.00010184 1.00010184 1.00010184 1.00010184 1.00010184 1.00019043
  1.00010184 1.00010184 1.00010184 1.00010184]
 [1.00010184 1.00010184 1.00010184 1.00010184 1.00010184 1.00019043
  1.00010184 1.00010184 1.00010184 1.00010184]]
Hyperspy: L=
[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 2. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]
Ganguli-Candes: S=
[[0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.99959858
  0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.
  0.         0.         0.         0.        ]]
Hyperspy: S=
[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]
Ganguli-Candes: L+S=
[[1.00010184 1.00010184 1.00010184 1.00010184 1.00010184 1.00019043
  1.00010184 1.00010184 1.00010184 1.00010184]
 [1.00010184 1.00010184 1.00010184 1.00010184 1.00010184 1.00019043
  1.00010184 1.00010184 1.00010184 1.00010184]
 [1.00031282 1.00031282 1.00031282 1.00031282 1.00031282 2.
  1.00031282 1.00031282 1.00031282 1.00031282]
 [1.00010184 1.00010184 1.00010184 1.00010184 1.00010184 1.00019043
  1.00010184 1.00010184 1.00010184 1.00010184]
 [1.00010184 1.00010184 1.00010184 1.00010184 1.00010184 1.00019043
  1.00010184 1.00010184 1.00010184 1.00010184]]
Hyperspy: L+S=
[[2. 2. 2. 2. 2. 2. 2. 2. 2. 2.]
 [2. 2. 2. 2. 2. 2. 2. 2. 2. 2.]
 [2. 2. 2. 2. 2. 3. 2. 2. 2. 2.]
 [2. 2. 2. 2. 2. 2. 2. 2. 2. 2.]
 [2. 2. 2. 2. 2. 2. 2. 2. 2. 2.]]
Done.
Hyperspy version 1.3
```

Just out of curiosity I also tried the same with larger dimensions and with the transpose of ```A```, but the result was equivalent, as it should be.

**Comment by francisco-dlp** [comment]
@Konrad0, HyperSpy's result is inconsistent and wrong. The issue is that the RPCA algorithm in HyperSpy performs internally a normalisation of the data that is typical in some domains but wrong in general. This issue is unfortunately not isolated: all the lazy decomposition algorithms suffer from the same malaise. Thanks for spotting and reporting the issue.

**Comment by tjof2** [comment]
@francisco-dlp can't you turn the normalisation off?

@Konrad0 fyi, the GoDec implementation came from https://kastnerkyle.github.io/posts/robust-matrix-decomposition/

**Comment by Konrad0** [comment]
@francisco-dlp, @tjof2 Thanks for the informative link and great to read that you're on to the issue already. Looking at the code, I guess the fix boils down to changing 
```python
    # Get min & max of data matrix for scaling
    X_max = np.max(X)
    X_min = np.min(X)
    X = (X - X_min) / X_max
    
    [...]
    
    # Rescale
    Xhat = (L * X_max) + X_min
    Ehat = (E * X_max) + X_min
    Ghat = (G * X_max) + X_min
```
to
```python
    # Get min & max of data matrix for scaling
    X_max = np.max(X)
    X_min = np.min(X)
    X = (X - X_min) / (X_max - X_min)
    
    [...]
    
    # Rescale
    Xhat = (L * (X_max - X_min)) + X_min
    Ehat = (E * (X_max - X_min))
    Ghat = (G * (X_max - X_min))
```

BTW, is there a reason why MinMax-scaling is used? Would standard scaling or some type of robust scaling work as well? Maybe someone has some experience with this (or even a reference).

Also, for the purpose of my work, the final SVD is not necessary, but I guess making it optional would break backward compatibility (and probably does not make the code a lot faster).

For ease of use, it might be nice to have a default argument for `rank`, like `rank = min(X.shape)`.

One point I couldn't understand from the paper is why we use `lambda1 = 1.0/np.sqrt(n)` as the default, where `n = min(X.shape)`. Our warning says "Threshold 'lambda1' is set to default: 1 / sqrt(nsamples)", but in my experience, the number of samples is usually much larger than the number of variables, so shouldn't it be `lambda1 = 1.0/np.sqrt(m)`? But this might be different in other fields of application, so I might be confused ;-)

**Comment by francisco-dlp** [comment]
@tjof2, you can turn it off for the lazy algorithms (it's on by default) but not for the algorithm discussed here. In addition, the lazy PCA algorithm uses sklearn's PCA that internally centres the data, something that cannot be disabled. I think that, for our purposes, internal scaling may be safe if needed (but is it needed?), but subtracting a number from the data may change the nature of the problem that we're trying to solve. That's why I have marked this as a bug (i.e. something that should fix in a patch release) rather than an API change (something that should only change in a major release).

@Konrad0, we can't disable the computation of the SVD step when calling the algorithm from ``Signal.decomposition`` because the purpose of this algorithm in HyperSpy is to perform a matrix decomposition, usually for dimensionality reduction purposes. Something that we could do though is to perform this step outside of the ``rpca_godec`` so function, so that the algorithm could be used on its own without computing the SVD.

automatic rank estimation is a missing feature in HyperSpy. As this value is critical in some cases and can have a high impact in the computation time, in the absence of a robust estimation method, we prefer to leave the user choose it rather than providing a default value that may be more often wrong than right.

**Comment by tjof2** [comment]
@Konrad0  MinMax scaling makes it easy to define the thresholds lambda1 and lambda2 as `1.0/something` - I found during testing that if your data is in the range [0, 10], then `10/something` works better... does that make sense? Basically it makes it easy to just use the default parameters, to make it accessible, rather than needing hyperparameter cross-validation etc.

Robust scaling (a la sklearn) would be fine also, but at the expense of needing to tune the thresholds a little more.

As for `1/sqrt(n)` vs `1/sqrt(m)` - it's somewhat arbitrary, you can always transpose your axes no? As it stands it makes sense from the point of view of typical HyperSpy data.

Also, I suspect the use of the term `1/sqrt(nsamples)` in the docstring may be a mistake, I know when @to266 converted this to use lazy algorithms we ended up transposing everything internally from how I originally wrote the code - the reference code I have locally has `1/sqrt(nfeatures)`, after all...

**Comment by Konrad0** [comment]
@tjof2 Thanks for the explanation. What confused me about the MinMax scaling is that it can change drastically due to outliers, a behavior we actually want to avoid with rPCA. That's why I was thinking of robust scaling, but since you tested it, there's no problem. Also, I guess you're right about the docstring, at this point in the code we may have already transposed the matrix, so it might be `1/sqrt(nsamples)` or `1/sqrt(nfeatures)`, depending on the dataset.

**Comment by tjof2** [comment]
@Konrad0 sure - outliers can change the scaling by a lot! But as long as your outliers are not > 1e10 I don't think you'll have any issues with floating point precision creeping in at all (I tested up to 1e8 on a low-rank matrix in the range [0,255] for example)

**Comment by ericpre** [comment]
Fixed in #1918.
